src/rust/aggregator/src/data/user_history.rs

```diff
@@ -0,0 +1,411 @@
+use anyhow::anyhow;
+use bigdecimal::{BigDecimal, Zero};
+use chrono::{DateTime, Duration, Utc};
+use sqlx::PgPool;
+
+use super::{Data, DataAggregationError, DataAggregationResult};
+
+#[derive(sqlx::Type, Debug)]
+#[sqlx(type_name = "order_status", rename_all = "lowercase")] // only for PostgreSQL to match a type definition
+pub enum OrderStatus {
+    Open,
+    Closed,
+    Cancelled,
+}
+
+#[derive(sqlx::Type, Debug)]
+#[sqlx(type_name = "order_type", rename_all = "lowercase")] // only for PostgreSQL to match a type definition
+pub enum OrderType {
+    Limit,
+    Market,
+    Swap,
+}
+
+pub struct UserHistory {
+    pool: PgPool,
+    last_indexed_timestamp: Option<DateTime<Utc>>,
+}
+
+impl UserHistory {
+    pub fn new(pool: PgPool) -> Self {
+        Self {
+            pool,
+            last_indexed_timestamp: None,
+        }
+    }
+}
+
+#[async_trait::async_trait]
+impl Data for UserHistory {
+    fn ready(&self) -> bool {
+        self.last_indexed_timestamp.is_none()
+            || self.last_indexed_timestamp.unwrap() + Duration::seconds(5) < Utc::now()
+    }
+
+    async fn process_and_save_historical_data(&mut self) -> DataAggregationResult {
+        self.process_and_save_internal().await
+    }
+
+    fn poll_interval(&self) -> Option<std::time::Duration> {
+        Some(std::time::Duration::from_secs(60 * 60))
+    }
+
+    async fn process_and_save_internal(&mut self) -> DataAggregationResult {
+        let fill_events = sqlx::query!(
+            r#"
+                SELECT * FROM fill_events
+                WHERE NOT EXISTS (
+                    SELECT * FROM aggregator.aggregated_events
+                    WHERE fill_events.txn_version = aggregated_events.txn_version
+                    AND fill_events.event_idx = aggregated_events.event_idx
+                )
+            "#,
+        )
+        .fetch_all(&self.pool)
+        .await
+        .map_err(|e| DataAggregationError::ProcessingError(anyhow!(e)))?;
+        let change_events = sqlx::query!(
```

Comment:
need to sort by the version and event idx so that we always aggregate the change size events in the order that they happen.

Reply:
The only case I could come up with where this could be an issue is if someone can close an order by changing size to 0. Is that a possibility ?